plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -509,6 +515,58 @@ inferTypeM (Error ann ty) = do
     checkKindM ann ty $ Type ()
     normalizeTypeM $ void ty
 
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_i ... s_n     s_i = [p_0 ... p_m]
+-- [check| G !- t_j : p_j]
+-- ----------------------------------------------------------------------------
+-- [infer| G !- constr resTy i t_0 ... t_n : vResTy]
+inferTypeM t@(Constr ann resTy i args) = do
+    vResTy <- normalizeTypeM $ void resTy
+
+    -- We don't know exactly what to expect, we only know what the i-th sum should look like, so we
+    -- assert that we should have some types in the sum up to there, and then the known product type.
+    let expectedType = TySOP () (replicate ((fromIntegral i) - 1) [dummyType] ++ [replicate (length args) dummyType])
+    case unNormalized vResTy of
+        TySOP _ vSTys -> case vSTys ^? wix i of
+            Just pTys -> case zipExact args pTys of
+                -- pTy is a sub-part of a normalized type, so normalized
+                Just ps -> for_ ps $ \(arg, pTy) -> checkTypeM ann arg (Normalized pTy)
+                -- the number of args does not match the number of types in the i'th SOP
+                -- alternative
+                Nothing -> throwing _TypeError (TypeMismatch ann (void t) expectedType vResTy)
+            -- result type does not contain an i'th sum alternative
+            Nothing -> throwing _TypeError (TypeMismatch ann (void t) expectedType vResTy)
+        -- result type is not a SOP type
+        _ -> throwing _TypeError (TypeMismatch ann (void t) expectedType vResTy)
+
+    pure vResTy
+
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_n     s_i = [p_i_0 ... p_i_m]
```

Comment:
Make it `sop s_0 ... s_i ... s_n` instead of `sop s_0 ... s_n` like in the rule above?
It's also `[p_i_0 ... p_i_m]` here and `[p_0 ... p_m]` there. I think I prefer the latter, since it has less clutter and it's easy to figure out the implicit `i` anyway.